File: templates/wasm-node-lua/examples/http_request.lua
-- HTTP Request Node — Demonstrates declaring HTTP permissions (Lua)
--
-- This example shows how to declare the "http" permission and use the
-- host import to make outbound HTTP requests from a Lua WASM node.
-- Copy this pattern into your node.lua when you need network access.

local sdk = require("sdk")

-- ============================================================================
-- Node definition — note addPermission("http")
-- ============================================================================

function get_node()
    local def = sdk.newNodeDefinition()
    def.name          = "http_get_request_lua"
    def.friendly_name = "HTTP GET Request (Lua)"
    def.description   = "Sends a GET request to a URL and reports the result"
    def.category      = "Network/HTTP"
    sdk.addPermission(def, "http")

    sdk.addPin(def, sdk.inputExec())
    sdk.addPin(def, sdk.withDefault(
        sdk.inputPin("url", "URL", "Target URL", sdk.DataType.String),
        '"https://httpbin.org/get"'
    ))
    sdk.addPin(def, sdk.withDefault(
        sdk.inputPin("headers_json", "Headers (JSON)", "Request headers as JSON", sdk.DataType.String),
        '"{}"'
    ))

    sdk.addPin(def, sdk.outputExec())
    sdk.addPin(def, sdk.outputPin("success", "Success", "Whether the HTTP call was accepted", sdk.DataType.Bool))

    return sdk.serializeDefinition(def)
end

function get_nodes()
    return "[" .. get_node() .. "]"
end

-- ============================================================================
-- Run handler — uses httpRequest host import directly
-- ============================================================================

function run_node(raw_json)
    local input = sdk.parseInput(raw_json)
    local ctx = sdk.newContext(input)

    local url = ctx:getString("url", "https://httpbin.org/get")
    local headers = ctx:getString("headers_json", "{}")

    sdk.logInfo("Sending GET request to " .. url)

    -- Method 0 = GET. The host checks the "http" capability.
    local resultCode = sdk.httpRequest(0, url, headers, "")
    local ok = resultCode ~= -1

    if ok then
        sdk.logInfo("HTTP capability granted — request dispatched")
    else
        sdk.logError("HTTP capability denied — is the 'http' permission declared?")
    end

    ctx:setOutput("success", ok and "true" or "false")

    local result = ctx:success()
    return sdk.serializeResult(result)
end

File: templates/wasm-node-assemblyscript/examples/http_request.ts
/**
 * HTTP Request Node — Demonstrates declaring HTTP permissions (AssemblyScript)
 *
 * This example shows how to declare the "http" permission and use the host
 * function to make outbound HTTP requests from an AssemblyScript WASM node.
 * Register this node in your examples.ts or index.ts entry point.
 */

import {
  DataType,
  NodeDefinition,
  PinDefinition,
  Context,
  ExecutionResult,
  FlowNode,
} from "@flow-like/wasm-sdk-assemblyscript/assembly/index";
import { httpRequest } from "@flow-like/wasm-sdk-assemblyscript/assembly/host";

// ============================================================================
// Node class — note `def.addPermission("http")`
// ============================================================================

export class HttpGetRequestNode extends FlowNode {
  define(): NodeDefinition {
    const def = new NodeDefinition();
    def.name = "http_get_request_as";
    def.friendly_name = "HTTP GET Request (AS)";
    def.description = "Sends a GET request to a URL and reports the result";
    def.category = "Network/HTTP";
    def.addPermission("http");

    def.addPin(
      PinDefinition.input("exec", "Execute", "Trigger execution", DataType.Exec)
    );
    def.addPin(
      PinDefinition.input("url", "URL", "Target URL", DataType.String)
        .withDefaultString("https://httpbin.org/get")
    );
    def.addPin(
      PinDefinition.input(
        "headers_json",
        "Headers (JSON)",
        "Request headers as JSON",
        DataType.String
      ).withDefaultString("{}")
    );
    def.addPin(
      PinDefinition.output("exec_out", "Done", "Fires after the request", DataType.Exec)
    );
    def.addPin(
      PinDefinition.output(
        "success",
        "Success",
        "Whether the HTTP call was accepted",
        DataType.Bool
      )
    );

    return def;
  }

  execute(ctx: Context): ExecutionResult {
    const url = ctx.getString("url");
    const headers = ctx.getString("headers_json");

    // Method 0 = GET.  The host checks the "http" capability.
    const ok = httpRequest(0, url, headers, "");

    ctx.setBool("success", ok);
    return ctx.success();
  }
}
